### PR TITLE
Fix the JS error "EventSource is not defined" caused by PaleMoon

### LINKED ISSUE #20572: `ReferenceError: EventSource is not defined` in Pale Moon
### Description

After updating to 1.17.0 with javascript enabled this time, after logging in I start getting errors about EventSource (see screenshot). (It doesn't happen when I'm logged out)

Developer console:
```
18:25:45.836 error { target: SharedWorker, isTrusted: true, message: "ReferenceError: EventSource is not defined", filename: "http://localhost:3000/assets/js/eventsource.sharedworker.js", lineno: 1, colno: 85, srcElement: SharedWorker, eventPhase: 0, bubbles: false, cancelable: true, defaultPrevented: false }  
index.js:58:1724
	ev/< http://localhost:3000/assets/js/index.js:58:1724
18:25:45.836 error { target: SharedWorker, isTrusted: true, message: "ReferenceError: EventSource is not defined", filename: "http://localhost:3000/assets/js/eventsource.sharedworker.js", lineno: 1, colno: 85, srcElement: SharedWorker, eventPhase: 0, bubbles: false, cancelable: true, defaultPrevented: false }  
index.js:37:15714
18:25:45.836 error { target: SharedWorker, isTrusted: true, message: "ReferenceError: EventSource is not defined", filename: "http://localhost:3000/assets/js/eventsource.sharedworker.js", lineno: 1, colno: 85, srcElement: SharedWorker, eventPhase: 0, bubbles: false, cancelable: true, defaultPrevented: false }  
index.js:58:1724
18:25:45.837 error { target: SharedWorker, isTrusted: true, message: "ReferenceError: EventSource is not defined", filename: "http://localhost:3000/assets/js/eventsource.sharedworker.js", lineno: 1, colno: 85, srcElement: SharedWorker, eventPhase: 0, bubbles: false, cancelable: true, defaultPrevented: false }  
index.js:37:15714
```

Actually, downgrading to 1.16.9 I still have the error, but only in the developer console not in the UI. As a workaround, I can add
```
[ui.notification]
EVENT_SOURCE_UPDATE_TIME = -1
```
to my `app.ini`, but `EVENT_SOURCE_UPDATE_TIME`'s documentation seems to suggest that the feature availability is auto-detected. (If I type `EventSource` into devconsole, I get back that it's a `function`, I don't know what gitea does to get it undefined...)

### Screenshots

This is running gitea locally, but I get the same result on try.gitea.org.
a

### Gitea Version

1.17.0

### Can you reproduce the bug on the Gitea demo site?

Yes

### Operating System

Linux

### Browser Version

Pale Moon 31.1.1

### PR DESCRIPTION
**Pull request by wxiaoguang**
# Background

Some browsers (like Pale Moon, Firefox<53) don't support EventSource in SharedWorkerGlobalScope. It's a bug https://stackoverflow.com/questions/25234493/server-sent-events-in-firefox-shared-web-workers .

Gitea's eventsource.sharedworker.js needs EventSource when doing "new Source(url)".

Before 1.17, even if JS error occurs, there is no prompt to end users, the code just didn't work.

Since 1.17, the UI will show JS errors, so if a users is using a such browser, they will see a JS error prompt. Thanks to our global JS error handler, we catch this problem now.

# Solution

The simple solution is to check whether there is a EventSource class in global (the self, aka SharedWorkerGlobalScope). If no EventSource class, just post a message back to the caller, in case the caller would like to use a fallback method to do its work.

~~Although it's feasible to make the worker's caller to handle the 'no-event-source' message to use a fallback method, I think it's out of the scope of this PR, because in such case, the real-time polling just never worked before, it's not worth to do the fix for the browser's bugs.~~

Now, the callers fall back to periodic poller if there is no EventSource

ps: there seems to be a backend bug, the worker can not get the "stop" event for stopwatch or the "1->0" change for notification when using EventSource, while other functions (stopwatch start event by EventSource/PeriodicPoller, stopwatch stop event by PeriodicPoller, notification by EventSource/PeriodicPoller) are all tested on my side.

Close #20572

### CODE REVIEW

**Comment by zeripath** [comment]
This PR does not do fallback to the poller function.

**Comment by wxiaoguang** [comment]
> This PR does not do fallback to the poller function.

Yup, described in the comment already

> Although it's feasible to make the worker's caller to handle the 'no-event-source' message to use a fallback method, I think it's out of the scope of this PR, because in such case, the real-time polling just never worked before, it's not worth to do the fix for the browser's bugs.

If there is one more people like to add the fallback back, I can do it.

**Comment by wxiaoguang** [comment]
@zeripath the poller is back

File: web_src/js/features/stopwatch.js [hunk 1]
```diff
@@ -26,28 +25,47 @@ export function initStopwatch() {
...
     worker.port.addEventListener('message', (event) => {
       if (!event.data || !event.data.type) {
-        console.error(event);
+        console.error('unknown worker message event', event);
         return;
       }
       if (event.data.type === 'stopwatches') {
         updateStopwatchData(JSON.parse(event.data.data));
+      } else if (event.data.type === 'no-event-source') {
+        console.error(`browser doesn't support EventSource, falling back to periodic poller`);
```
Comment (silverwind) [inline web_src/js/features/stopwatch.js]: It's not a critical error if there's a fallback, I wouldn't log it. Generally we should keep console clean for actual critical errors.

**Review by zeripath — APPROVED** [review]

File: web_src/js/features/stopwatch.js (hunk 1 above)
Comment (wxiaoguang) [reply web_src/js/features/stopwatch.js]:
```suggestion
        // browser doesn't support EventSource, falling back to periodic poller
```

File: web_src/js/features/notification.js [hunk 2]
```diff
@@ -49,28 +46,41 @@ export function initNotificationCount() {
...
     worker.port.addEventListener('message', (event) => {
       if (!event.data || !event.data.type) {
-        console.error(event);
+        console.error('unknown worker message event', event);
         return;
       }
       if (event.data.type === 'notification-count') {
         const _promise = receiveUpdateCount(event.data);
+      } else if (event.data.type === 'no-event-source') {
+        console.error(`browser doesn't support EventSource, falling back to periodic poller`);
```
Comment (wxiaoguang) [inline web_src/js/features/notification.js]:
```suggestion
        // browser doesn't support EventSource, falling back to periodic poller
```

File: web_src/js/features/notification.js [hunk 3]
```diff
@@ -28,14 +28,11 @@ async function receiveUpdateCount(event) {
   try {
     const data = JSON.parse(event.data);
 
-    const notificationCount = document.querySelector('.notification_count');
-    if (data.Count > 0) {
-      notificationCount.classList.remove('hidden');
-    } else {
-      notificationCount.classList.add('hidden');
+    const notificationCounts = document.querySelectorAll('.notification_count');
+    for (const count of notificationCounts) {
```
Comment (silverwind) [inline web_src/js/features/notification.js]:
```suggestion
    for (const count of document.querySelectorAll('.notification_count')) {
```

File: web_src/js/features/stopwatch.js [hunk 4]
```diff
@@ -26,28 +25,47 @@ export function initStopwatch() {
...
+    updateStopwatchTime(currSeconds);
+  }
+
+  let usingPeriodicPoller = false;
+  const startPeriodicPoller = (timeout) => {
+    if (timeout <= 0 || !Number.isFinite(timeout)) return;
+    usingPeriodicPoller = true;
+    setTimeout(async () => {
+      await updateStopwatchWithCallback(startPeriodicPoller, timeout);
+    }, timeout);
```
Comment (silverwind) [inline web_src/js/features/stopwatch.js]:
```suggestion
    setTimeout(() => updateStopwatchWithCallback(startPeriodicPoller, timeout), timeout);
```

Seems useless to `await` if you do nothing after it.

**Review by silverwind — APPROVED** [review]
LGTM after above suggestions.

**Review by lafriks — APPROVED** [review]

**Comment by lunny** [comment]
Maybe it's valuable to backport to v1.17 when possible.

**Comment by wxiaoguang** [comment]
I would like to wait for feedbacks from users (https://github.com/go-gitea/gitea/issues/20572#issuecomment-1204621780) first.

**Comment by wxiaoguang** [comment]
The main branch has been confirmed by https://github.com/go-gitea/gitea/issues/20572#issuecomment-1204958859

The backport for 1.17 is in https://github.com/go-gitea/gitea/pull/20663

**Comment by wolfbeast** [comment]
I just upgraded to 1.17.1 and this error doesn't seem to be fixed on Pale Moon.
```
JavaScript error: ReferenceError: EventSource is not defined (https://repo.palemoon.org/assets/js/eventsource.sharedworker.js @ 1:0). Open browser console to see more details.
```

**Comment by wxiaoguang** [comment]
Cache problem (I have read other comments, just put a comment for it here)

**Comment by silverwind** [comment]
Cache issue will be solved with https://github.com/go-gitea/gitea/pull/20813.

**Comment by wolfbeast** [comment]
> Cache issue will be solved with #20813.

Awesome! It bit me because even a hard refresh didn't actually get me where I wanted to. Only a full browser cache purge did the trick (may even need a browser restart, depending on configuration!)

**Pull request merged**